src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,399 @@
+package pfsdb
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/coredb"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+)
+
+const (
+	getRepoAndBranches = "SELECT repo.id, repo.name, repo.type, repo.project_id, " +
+		"repo.description, array_agg(branch.proto) AS branches, repo.created_at, repo.updated_at FROM pfs.repos repo " +
+		"LEFT JOIN core.projects project ON repo.project_id = project.id " +
+		"LEFT JOIN collections.branches branch ON project.name || '/' || repo.name || '.' || repo.type = branch.idx_repo "
+	noBranches = "{NULL}"
+)
+
+// ErrRepoNotFound is returned by GetRepo() when a repo is not found in postgres.
+type ErrRepoNotFound struct {
+	Project string
+	Name    string
+	Type    string
+	ID      pachsql.ID
+}
+
+// Error satisfies the error interface.
+func (err ErrRepoNotFound) Error() string {
+	if id := err.ID; id != 0 {
+		return fmt.Sprintf("repo id=%d not found", id)
+	}
+	return fmt.Sprintf("repo %s/%s.%s not found", err.Project, err.Name, err.Type)
+}
+
+func (err ErrRepoNotFound) Is(other error) bool {
```

Comment:
I'm a  bit confused by the intent of this function. And how we check for errors here in general.

This `ErrRepoNotFound.Is` is supposed to customize the behaviour of `errors.Is`, but right now it's basically returning true for any error that has the type `ErrRepoNotFound`. However, the point of `errors.Is` is to be value aware.

For example
```
err1 := ErrRepoNotFound{Name: "repo1"}
err2 := ErrRepoNotFound{Name: "repo2"}

// without customization
errors.Is(err1, err2) // returns false
// however
err1.Is(err2) // returns true
```

Based how you use `ErrRepoNotFound.Is` in tests, it seems like you care about the value after all. I think there's a simpler way to do this with native Go library

```
err := ErrRepoNotFound{Name: "repo1"}
err2 := ErrRepoNotFound{Name: "repo1"}
err3 := ErrRepoNotFound{Name: "repo2"}

// to compare two errors by value
fmt.Println(errors.Is(err, err2))               // false
fmt.Println(errors.Is(err, err3))               // false
// to check an error's type
fmt.Println(errors.As(err, &ErrRepoNotFound{})) // true
```

Lastly, a further argument for using `errors.Is` and `errors.As` is to handle wrapped errors.
For example
```
err := fmt.Errorf("adding context: %w", ErrRepoNotFound{Name: "repo1"})
_, ok := err.(ErrRepoNotFound) // ok == false
// but
errors.As(err, &ErrRepoNotFound{}) // returns true
```

Reply:
I think I got this from a suggestion on projects and I just copy and pasted the same structure